Mathlib/Combinatorics/HalesJewett.lean

```diff
@@ -0,0 +1,384 @@
+`{ (a, x, x, b, x) | x : α }`. See `Combinatorics.Line` for a precise general definition. The
+Hales-Jewett theorem states that for any fixed finite types `α` and `κ`, there exists a (potentially
+huge) finite type `ι` such that whenever `ι → α` is `κ`-colored (i.e. for any coloring
+`C : (ι → α) → κ`), there exists a monochromatic line. We prove the Hales-Jewett theorem using
+the idea of *color focusing* and a *product argument*. See the proof of
+`Combinatorics.Line.exists_mono_in_high_dimension'` for details.
+
+The version of Van der Waerden's theorem in this file states that whenever a commutative monoid `M`
+is finitely colored and `S` is a finite subset, there exists a monochromatic homothetic copy of `S`.
+This follows from the Hales-Jewett theorem by considering the map `(ι → S) → M` sending `v`
+to `∑ i : ι, v i`, which sends a combinatorial line to a homothetic copy of `S`.
+
+## Main results
+
+- `Combinatorics.Line.exists_mono_in_high_dimension`: the Hales-Jewett theorem.
+- `Combinatorics.exists_mono_homothetic_copy`: a generalization of Van der Waerden's theorem.
+
+## Implementation details
+
+For convenience, we work directly with finite types instead of natural numbers. That is, we write
+`α, ι, κ` for (finite) types where one might traditionally use natural numbers `n, H, c`. This
+allows us to work directly with `α`, `Option α`, `(ι → α) → κ`, and `ι ⊕ ι'` instead of `Fin n`,
+`Fin (n+1)`, `Fin (c^(n^H))`, and `Fin (H + H')`.
+
+## Todo
+
+- Prove a finitary version of Van der Waerden's theorem (either by compactness or by modifying the
+current proof).
+
+- One could reformulate the proof of Hales-Jewett to give explicit upper bounds on the number of
+coordinates needed.
+
+## Tags
+
+combinatorial line, Ramsey theory, arithmetic progession
+
+### References
+
+* https://en.wikipedia.org/wiki/Hales%E2%80%93Jewett_theorem
+
+-/
+
+
+open Classical
+
+-- Porting note: commented out the next line
+-- open BigOperators
+
+universe u v
+
+namespace Combinatorics
+
+/-- The type of combinatorial lines. A line `l : Line α ι` in the hypercube `ι → α` defines a
+function `α → ι → α` from `α` to the hypercube, such that for each coordinate `i : ι`, the function
+`fun x ↦ l x i` is either `id` or constant. We require lines to be nontrivial in the sense that
+`fun x ↦ l x i` is `id` for at least one `i`.
+
+Formally, a line is represented by the function `l.idxFun : ι → Option α` which says whether
+`fun x ↦ l x i` is `id` (corresponding to `l.idxFun i = none`) or constantly `y` (corresponding to
+`l.idxFun i = some y`).
+
+When `α` has size `1` there can be many elements of `Line α ι` defining the same function. -/
+structure Line (α ι : Type _) where
+  /-- The function `l` underlying a combinatorial line; For each `i : ι`, either `fun x ↦ l x i` is
+   the identity function on `α` (corresponding to `l.idxFun i = none`) or constantly `y`
+   (corresponding to `l.idxFun i = some y`). -/
```

Comment:
This explanation could use some more work. It seems confusing to call it "The function underlying a combinatorial line", since this is _not_ the CoeFun from lines. Otherwise the new docstrings look quite good!

Reply:
You're welcome to push a fix -- since you're the original author of the file, you're probably in a better position to make this kind of change than the people porting it!

Reply:
I've marked this PR as awaiting-author until somebody write a better explanation.

Reply:
Can you make a suggestion for a better doc-string, @dwarn?

Reply:
If you make a suggestion what to call it instead of "The function underlying a combinatorial line", I'll gladdy edit the file, and get this merged,  @dwarn

Reply:
I just pushed a small change. Sorry to hold this up for so long!

Reply:
Thanks, great!